### ISSUE TITLE
DL4J: Benchmarking and Performance Investigations

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#5288]
This issue is intended to summarize some performance results and plan next steps for optimizations in DL4J.

Summary:
* Benchmarks show that 1.0.0-beta is considerably faster than 1.0.0-alpha, but slower than 0.9.1
    * vs. 0.9.1, 1.0.0-beta is about 14% slower (Compgraph Resnet V1, minibatch 32, Linux) to 24.1% slower (MLN AlexNet, minibatch 32, Linux)
* Windows is slower: somewhere on the order of 5-25% vs. Linux (both with CuDNN)
* Memory use is considerably lower (better) in 1.0.0-beta than prior versions

Benchmark results (currently private to Skymind devs only):
https://docs.google.com/spreadsheets/d/1uMzZ64ZMW8QkgXN-Wx-C1wcWj7CJU7g3V-ShUkN1znk/edit?usp=sharing
But can be replicated using: https://github.com/deeplearning4j/dl4j-benchmark/tree/master/scripts

Other reports:
* https://github.com/deeplearning4j/deeplearning4j/issues/5286 - reporting training 2.2x slower with 1.0.0-beta vs. 0.9.1, using CuDNN
    * This doesn't match what I'm seeing in benchmarks - I need to run the exact configuration to confirm...

YourKit Snapshots: TBD

**Comment by DrChainsaw** [comment]
Adding here as well since my comment in #5286 might be lost: Factor of 2.2 is with Nesterovs updater. Changing from Nesterovs to Adam in 1.0.0 beta changes runtime to about 1.2 of 0.9.1. 

In 0.9.1 there is no difference between Adam and Nesterovs.

**Comment by AlexDBlack** [comment]
@DrChainsaw thanks. That's unexpected... I don't recall any significant changes to updaters between 0.9.1 and 1.0.0-beta. I'll definitely take a look and benchmark that too :)

**Comment by AlexDBlack** [comment]
@DrChainsaw Yep, definitely something up with Nesterovs... I've updated original issue with updater benchmark results.

**Comment by DrChainsaw** [comment]
Hi,

Sorry if this counts as necroing. I have been running my dummy benchmark from #5286 on snapshots every now and then and seen performance getting better and better up to about 50% shorter training times compared to 9.1. 

Just a couple of days ago when I ran it performance was drastically worse. Training was back at about 20% slower while inference was something like 200% slower.

**Comment by agibsonccc** [comment]
@DrChainsaw did you pull the latest snapshots and try it with alex's fixes? If so, could you post your logs the same as before? Descriptions like this don't really help much.

**Comment by AlexDBlack** [comment]
@DrChainsaw yeah, pull latest snapshots and double check. Some numbers on what you're seeing (as well as specifics - hardware, what you ran, when, etc) would be helpful too.

**Comment by DrChainsaw** [comment]
Sorry for not giving details. I wanted to give it a few days first, forgot about it and remembered it when I was out of town with only limited remote access using my phone. Hardware and setup is same as in #5286.

Anyways, performance seems much better on today's snapshot and it beats 9.1 with a large margin. I think I saw even faster training times before though, but it might as well be a change I made to the config and forgot about.

I'll look into it some more when I get the opportunity and repost here if I find something.

Sorry again for low quality report. I'm just trying to help out. For all I know, you might not even been aware that performance had changed for the better and the report at least gave that info along with a rough timeframe as well.

**Comment by DrChainsaw** [comment]
Sorry again if I keep beating a dead horse here. Feel free to tell me to stop pestering you about this :)

Anyways, I could not find any feasible way to go back to old snapshots in order to investigate my previous comment.

One thing I did manage to look into was that it seems like architectures with many layers took a big performance hit when WorkspaceMode.SINGLE was deprecated. For example, the ResNet50 model from the zoo runs about 100% slower in latest snapshot compared to 0.9.1 when SINGLE workspace mode is used. Maybe not so strange, but it might give a hint on further opimizations.

Test program I used:
https://gist.github.com/DrChainsaw/dad0ce89b9aa379d1c1ee517398112a5

Output for SNAPSHOT:
https://gist.github.com/DrChainsaw/ec94bdc580bc5fbed09d046328bc717f

Output for 0.9.1 with wsMode SINGLE:
 https://gist.github.com/DrChainsaw/c7c6f143a91a956c11486c1a1f4f7971

Output for 0.9.1 with wsMode SEPARATE:
https://gist.github.com/DrChainsaw/21bb489edf0d724c736285e24417dcc8

When I ran nvprof the outstanding difference I could spot was about 30 times more calls to cudaStreamSynchronize and about 2 times more calls to cudaEventSynchronize which I guess makes some sense. 

Sorry if this is stating the obvious. I guess you can remove this post to reduce clutter.

**Comment by agibsonccc** [comment]
Closing — this issue is over 7 years old with no recent activity. The codebase has undergone a major rewrite. If this is still relevant, please open a new issue with current reproduction steps.
